src/obj_char_transfer.c

```diff
@@ -0,0 +1,872 @@
+        if (*blockNumMain == -1) {
+            GF_ASSERT(FALSE);
+            return FALSE;
+        }
+        *blockNumMain = ObjCharTransfer_GetBlockMemOffset(*blockNumMain, sObjCharTransferTasksManager->blockSizeMain);
+    }
+    if (vram & NNS_G2D_VRAM_TYPE_2DSUB) {
+        *freeSpaceSub = ObjCharTransfer_AlignToBlock(size, sObjCharTransferTasksManager->blockSizeSub, TRUE);
+        u32 numBlocksUsed = ObjCharTransfer_CalcBlockNumLimit(*freeSpaceSub, sObjCharTransferTasksManager->blockSizeSub);
+        *blockNumSub = ObjCharTransfer_FindAvailableTransferBlock(numBlocksUsed, sObjCharTransferTasksManager->blockBufSub);
+        if (*blockNumSub == -1) {
+            GF_ASSERT(FALSE);
+            return FALSE;
+        }
+        *blockNumSub = ObjCharTransfer_GetBlockMemOffset(*blockNumSub, sObjCharTransferTasksManager->blockSizeSub);
+    }
+    return TRUE;
+}
+
+static void ObjCharTransferTask_UpdateBaseAddrs(ObjCharTransferTask *task, u32 offsetMain, u32 offsetSub) {
+    if (task->vram & NNS_G2D_VRAM_TYPE_2DMAIN) {
+        task->baseAddrMain = offsetMain + sObjCharTransferTasksManager->freeSizeMain;
+    }
+    if (task->vram & NNS_G2D_VRAM_TYPE_2DSUB) {
+        task->baseAddrSub = offsetSub + sObjCharTransferTasksManager->freeSizeSub;
+    }
+}
+
+static void reserveTransferBlocksByVramOffsetAndSize(NNS_G2D_VRAM_TYPE vram, u32 offsetMain, u32 offsetSub, u32 sizeMain, u32 sizeSub) {
+    if (vram & NNS_G2D_VRAM_TYPE_2DMAIN) {
+        u32 numBlocks = ObjCharTransfer_CalcBlockNumLimit(sizeMain, sObjCharTransferTasksManager->blockSizeMain);
+        u32 blockStart = ObjCharTransfer_CalcBlockNumLimit(offsetMain, sObjCharTransferTasksManager->blockSizeMain);
+        ObjCharTransfer_ReserveTransferBlocks(blockStart, numBlocks, sObjCharTransferTasksManager->blockBufMain);
+    }
+    if (vram & NNS_G2D_VRAM_TYPE_2DSUB) {
+        u32 numBlocks = ObjCharTransfer_CalcBlockNumLimit(sizeSub, sObjCharTransferTasksManager->blockSizeSub);
+        u32 blockStart = ObjCharTransfer_CalcBlockNumLimit(offsetSub, sObjCharTransferTasksManager->blockSizeSub);
+        ObjCharTransfer_ReserveTransferBlocks(blockStart, numBlocks, sObjCharTransferTasksManager->blockBufSub);
+    }
+}
+
+static void ObjCharTransferTask_G2dLoadImageMapping(ObjCharTransferTask *task) {
+    NNS_G2dInitImageProxy(&task->imageProxy);
+    if (task->vram != NNS_G2D_VRAM_TYPE_BOTH) {
+        ObjCharTransferTask_G2dLoadImageMappingByScreen(task, task->vram);
+    } else {
+        ObjCharTransferTask_G2dLoadImageMappingByScreen(task, NNS_G2D_VRAM_TYPE_2DMAIN);
+        ObjCharTransferTask_G2dLoadImageMappingByScreen(task, NNS_G2D_VRAM_TYPE_2DSUB);
+    }
+}
+
+static void ObjCharTransferTask_G2dLoadImageMappingByScreen(ObjCharTransferTask *task, NNS_G2D_VRAM_TYPE vram) {
+    int mappingType = task->shouldUpdateMappingTypeFromHW ? ObjCharTransferTask_SetMappingTypeFromHW(task, vram) : vram;
```

Comment:
should this one really be an int, I know the vram assignment is weird here though

Reply:
the exact type this should be is incredibly unclear, that's gamefreak's fault